### ISSUE TITLE
Mock only some specific functions in the same file (partial mocking)

### ISSUE THREAD

**Issue by parmi93** [ThrowTheSwitch/Ceedling#936]
I know this topic has been discussed at length, and several workaround have already been presented, but none of the existing solutions satisfy me fully, and I think Ceedling can do better(?) (although I am not sure that my proposal is feasible).

The workaround that is usually proposed is to move the static functions into private helper file and include it in the main file, so the helper file can be tested separately, and then use the "mocked version" of the helper file when you want to test the main file.
But sometimes this is not possible for example when working with legacy code, also if a helper function calls another helper function I have to separate these functions into different files (which I don't like at all), and in any case sometimes I want to keep my static functions in the main file because it makes sense according to the design of my application.

I would like to propose the following solution:
```c
// my_file.c
static int prv_a, prv_b, prv_c;
static int prv_n;

static void prv_foo_N(int n)
{
    prv_n = n;
}

static void prv_foo_AB(int a, int b)
{
    prv_a = a;
    prv_b = b;
    prv_foo_N(a + b);
}

void bar(int a, int b, int c)
{
    prv_c = c;
    prv_foo_AB(a, b);
}
```

```c
// test_bar_fn.c

/* Tell Ceedling that I want the file "my_file.c" to be included directly in
 * this test file.
 * So Ceedling could clone the file "my_file.c" into a file with the following
 * name "build/test/original_src/src_<this_test_file_name>.c", and then replace
 * this macro with: 
 * #include "build/test/original_src/src_test_bar_fn.c" (yes, .c) */
INCLUDE_SOURCE("my_file.c");

/* Tell Ceedling to literally remove the prv_foo_AB() function from the 
 * "build/test/original_src/src_test_bar_fn.c" file, and use the "mocked version"
 * of it.
 * The "mock version" of prv_foo_AB() could be created in a file with the 
 * following name "build/test/partial_mocks/mock_my_file_c_<function_name>.c"
 * which contain only the "mocked version" of prv_foo_AB() function, and then
 * replace this macro with:
 * #include "build/test/partial_mocks/mock_my_file_c_prv_foo_AB.c" (yes, .c) */
MOCK_FUNCTION("prv_foo_AB");

void test_bar()
{
    // Function defined in build/test/partial_mocks/mock_my_file_c_prv_foo_AB.c
    prv_foo_AB_Expect(2, 8, 99);

    // Function defined in build/test/original_src/src_test_bar_fn.c
    bar(2, 9, 99);
    
    TEST_ASSERT_EQUAL_INT(99, prv_n);
}
```

```c
// test_prv_foo_AB_fn.c

/* Tell Ceedling that I want the file "my_file.c" to be included directly in
 * this test file.
 * So Ceedling could clone the file "my_file.c" into a file with the following
 * name "build/test/original_src/src_<this_test_file_name>.c", and then replace
 * this macro with: 
 * #include "build/test/original_src/src_test_prv_foo_AB_fn.c" (yes, .c) */
INCLUDE_SOURCE("my_file.c");

/* Tell Ceedling to literally remove the prv_foo_N() function from the
 * "build/test/original_src/src_test_prv_foo_AB_fn.c" file, and use the
 * "mocked version" of it.
 * The "mock version" of prv_foo_N() could be created in a file with the
 * following name "build/test/partial_mocks/mock_my_file_c_<function_name>.c"
 * which contain only the "mocked version" of prv_foo_N() function, and then
 * replace this macro with:
 * #include "build/test/partial_mocks/mock_my_file_c_prv_foo_N.c" (yes, .c) */
MOCK_FUNCTION("prv_foo_N");

void test_prv_foo_AB()
{
    // Function defined in build/test/partial_mocks/mock_my_file_c_prv_foo_N.c
    prv_foo_N_Expect(10);

    // Function defined in build/test/original_src/src_test_prv_foo_AB_fn.c
    prv_foo_AB(2, 8);

    TEST_ASSERT_EQUAL_INT(2, prv_a);
    TEST_ASSERT_EQUAL_INT(8, prv_b);
}
```

```c
// test_prv_foo_N_fn.c

/* In this specific case this macro could simply be replaced with:
 * #include "my_file.c" (yes, .c), so including the original source file 
 * without cloning it, because no MOCK_FUNCTION() marco is used in this test
 * file. */
INCLUDE_SOURCE("my_file.c")

void test_prv_foo_N()
{
    // Function defined in my_file.c
    prv_foo_N(10); 

    TEST_ASSERT_EQUAL_INT(10, prv_n);
}
```

Basically `INCLUDE_SOURCE("file_name.c")` tells Ceedling to copy the file `file_name.c` into `build/test/original_src/src_<file_name>.c`, so the `original_src` folder should contain a copy of the original source files from which only the functions specified by the `MOCK_FUNCTION("function_name")` macros need to be removed.
This eliminates function redefinition errors during compilation and linking.

Would it be possible to implement something like this in Ceedling?

**Comment by mvandervoord** [comment]
:) What you are describing here is very close to how `scalpel` works, which is a new CMock feature. It's currently under development, but likely going to be the release after the soon-to-be-released version.

**Comment by parmi93** [comment]
This is great news!!
Could you please link me something about this `scalpel`, I can't find anything on Google about it.

**Comment by mvandervoord** [comment]
Sadly at the moment it exists only as a collection of files on my personal repo. It's staged to get added to the next release when some things are worked out.

**Comment by parmi93** [comment]
ok thanks for the info.
As far as you know, are there any unit test frameworks out there that support this feature?
looking around it seems like no one supports such a thing

**Comment by mvandervoord** [comment]
Not that I am aware of. As far as I know, we'd be the first... and that makes me excited to work on it. :)

**Comment by parmi93** [comment]
In my opinion this will be one of the features of CMock/Ceedling that will make it stand out from the crowd.

**Comment by mkarlesky** [comment]
Related issues to collocate in a single issue: #631 #604

**Comment by radoslav06** [comment]
I like the idea very much!
It will be also useful to have a possibility to mock all the static functions at once. Something like:
```C
MOCK_STATIC_FUNCTIONS();
```

**Comment by jmrubillon** [comment]
For legacy codebases this feature will be essential as re-organising an entire codebase to introduce a testing framework is unlikely to fly due to the risk of having to revalidate the entirety of the software.
I'm in the middle of doing just that and failing miserably to get a test running on a "module" that has a lot of static functions :(

**Comment by parmi93** [comment]
It's a smart way to mock `static` functions, thanks for sharing, same approach can be used to mock `inline` functions.

That said, I’d prefer not to clutter the codebase with test-specific code, we aim to keep it focused solely on production logic. Hopefully, CMock/Ceedling will eventually support this functionality natively.

Example project based on your approach by adding the mocking of `inline` function:
[ceedling_static_and_inline_mocking.zip](https://github.com/user-attachments/files/21986407/ceedling_static_and_inline_mocking.zip)

**Comment by ciocari** [comment]
Any due date for this feature?

**Comment by EmileBV** [comment]
Hi @mkarlesky, do you have any update on the progress of this feature? My team and I keep finding situations where this exact features might be the solution we are looking for. I don't want to rush anyone, just letting you know there's a lot of interest in this.

**Comment by mkarlesky** [comment]
@EmileBV Indeed I do. Apart from some final polish work and preparing the next release of Ceedling with fixes, etc. the new Partials feature is complete. We just finished the documentation. When the next prerelease is available I'll message you and point you to the docs.

Because this feature depends heavily on lexing C code for its magic, we are entirely expecting to find failures in complex and legacy C code as we encounter it.

p.s. We are also finishing up a brand new web-based, comprehensive, searchable documentation site for Ceedling. Partials is already part of that site (not yet published).

**Comment by EmileBV** [comment]
Nice, thanks for the information!

**Comment by mkarlesky** [comment]
@EmileBV The latest [pre-release of Ceedling with Partials support](https://github.com/ThrowTheSwitch/Ceedling/releases/tag/1.1.0-6f84d88) is available. The notes in the pre-release also link to Ceedling’s new documentation system including the Partials subsection.

Of course, Partials are very capable but also very new. We're only just now entering into stress testing this new, sophisticated set of features. Complicated code and special project setups are sure to reveal bugs and gaps.

**Comment by EmileBV** [comment]
Thanks for the update, I tried installing the pre-release gem but got an error despite it saying it completed installation. Then trying to run ceedling on a test yields the following error:
```
stderr:
🧨 EXCEPTION: undefined method '=~' for an instance of UserInclude
🌱 Ceedling could not complete operations because of errors
```

During installation, I got the following output:
[ceedling_install_error.txt](https://github.com/user-attachments/files/27415638/ceedling_install_error.txt)

No need to debug here, will wait for the full release, but I thought it might help to share these.

**Comment by mkarlesky** [comment]
@EmileBV This is helpful installation debugging. I suspect this is a Ruby version issue. What version of Ruby are you running?

**Comment by EmileBV** [comment]
@mkarlesky using Ruby 3.4.8 (exact version: `ruby-3.4.8-29.fc43.src.rpm`). I think it's the latest available on Fedora 43 without forcing a specific version through additional package repositories.

**Comment by mkarlesky** [comment]
@EmileBV Ah ha. The issue is in the log file. We've seen this once before. The installation reports success because the Ceedling gem fully installs. What is failing is RDoc trying to generate documentation from that install afterwards. RDoc must be able to parse Ruby in order to do its job. Older versions of RDoc sometimes fall behind or have bugs in their parsing. That's what is happening here. We haven't dug in to figure out the specific RDoc problem and if it's fixed in a more recent version. And, Ceedling's comments are not set up for RDoc anyhow.

You have two options:

1. Upgrade RDoc and hope it fixes the problem.
1. Add ` --no-document` to your Ceedling gem install command line.

**Comment by EmileBV** [comment]
@mkarlesky Hi, sorry for the delayed response. Tried both suggestions, pretty much the same result. Rdoc throws some errors during installation and then trying to run Ceedling on tests that work fine with 1.0.1 have the error I sent above (the `=~` method not found thing). Maybe there's new configuration to add to the project.yml file?

If that's the case I'll wait for the full 1.1.0 release for all the documentation to be ready so we can migrate our project to the new version.

**Comment by mkarlesky** [comment]
@EmileBV Looking closer at what you reported, there's two issues. One is purely the RDoc problem that suggests installation is not working, when it is, in fact, working. The other issue with `=~` is a legitimate bug relating to internal changes to better support includes handling. So far I have not been able to find the line of code in the project that is failing for you. Would you mind running your tests again with `--verbosity debug` (or `-v 5` in short form)?

**Comment by EmileBV** [comment]
@mkarlesky Sure thing, here's the output with `-v 5` on a small test file:
(I modified the path in the first few lines to hide a project name but otherwise the file is unmodified from the ceedling output)
[ceedling_error_log.txt](https://github.com/user-attachments/files/27765611/ceedling_error_log.txt)

**Comment by mkarlesky** [comment]
@EmileBV This is quite curious. I'm not sure what to make of the debug trace in the error log quite yet. But, this appears to be a different error than the one you first reported? Or, am I not understanding something?

```
stderr:
🧨 EXCEPTION: undefined method '=~' for an instance of UserInclude
🌱 Ceedling could not complete operations because of errors
```

**Comment by EmileBV** [comment]
@mkarlesky Sorry for the missing context, the log file is everything printed with `-v 5` before that ceedling error gets printed.

The command I executed to get that file is:
```
ceedling -v 5 test:median_filter > ceedling_error_log.txt
```
The output you quoted gets printed to the console without going to the file after it's done executing.

I hope this clears out any confusion.

**Comment by mkarlesky** [comment]
@EmileBV Okay. It all makes sense now! The problem was that a new internal class `UserInclude` ended up in a Rake `FileList`. The FileList attempted to do some matching on `UserInclude`, but the class lacked definitions for matching operators. Because a missing operator is towards the top of the exception tree it gets handled in a rarely used top-level exception handling block that cannot route that failure into the logging. The reason we've not seen this already is because you're `#include`ing a .c file in your test. Ceedling handles this fine, but that .c include triggered a special handling code path for gathering those .c files that we hadn't exercised in testing.

It should be all fixed now. Any pre-releases more recent than `1.1.0-2751b0c` (May 13, 2026 EDT) will have the fixes.

Thank you for the report. This was very helpful (if quite confusing at first!).

Incidentally, you may want to look at using `TEST_SOURCE_FILE()` instead of directly `#include`ing a .c file in your test. Here's [more on this](https://throwtheswitch.github.io/Ceedling/dev/testing-guide/build-directives/) in Ceedling's not-yet-officially published documentation site.

**Comment by EmileBV** [comment]
@mkarlesky Glad I could help. We've had to directly include the .c file in our tests to access the static functions. We'll change to the new system when it's officially released, I'm happy to learn there's now a more elegant way to do this!

**Comment by EmileBV** [comment]
@mkarlesky Hi, I tested using the latest pre-release gem (1.1.0-9b85ef0) and I'm getting a new error:
```
stderr:
🧨 EXCEPTION: undefined method 'pathmap' for an instance of UserInclude
🌱 Ceedling could not complete operations because of errors
```

Here's the output that got printed before: 
- Command: `ceedling -v 5 test:median_filter > ceedling_error_log.txt`
- Output: [ceedling_error_log.txt](https://github.com/user-attachments/files/28018744/ceedling_error_log.txt)

The includes section in that test file is like this:

```c
#include "unity.h"
#ifndef TEST_EDITING
/* Add mocks here. */
#include "mock_utility.h"
#endif  // !TEST_EDITING

/* Add regular includes here. */
#include "median_filter.c"
#include "median_filter.h"
```

The only file included by the tested files is `utility.h`, which includes system header files and some helper `typedef` and macros. When I run our "reference" test file, which is only a call to `TEST_PASS()`, it completes without issues.

Let me know if you need more information or if this is expected with the current state of the pre-release.

**Comment by mkarlesky** [comment]
@EmileBV Thank you for this. It's basically just a flavor of the same problem we fixed last go around. The triggering circumstance is `#include`ing a .c file in test cases. We just need to expand the system test suite to exercise this scenario. There may be additional failures like this one. I don't want you having to popcorn report all such bugs. We should just fully exercise this circumstance in testing. Let me do that and find and fix remaining issues triggered by `#include`ing .c files. I'll get back to you soon.

**Comment by EmileBV** [comment]
@mkarlesky Alright thanks! Also just letting you know I tried using the `TEST_SOURCE_FILE()` macro to avoid including the .c file, but got compilation errors as if nothing from the specified source file was found. Maybe I used it wrong or I'm missing something else in my test file? I followed the instructions in the docs you linked.

**Comment by mkarlesky** [comment]
@EmileBV My apologies on `TEST_SOURCE_FILE()`. I pointed to it before I knew you were testing `static` functions. The behavior you describe is expected when using it with static functions. It only serves to tell Ceedling to compile and link the named file with the test executable. Before the advent of Partials, your only option was to `#include` .c files so static functions were in the same translation unit and accessible. Partials provides a superior option. We just need to finish it up and ship it.

I did find and fix the problem we've been discussing and have a system test successfully exercising your testing scenario.

**Comment by mkarlesky** [comment]
@EmileBV The latest [pre-release of Ceedling](https://github.com/ThrowTheSwitch/Ceedling/releases/tag/1.1.0-7343c1c) should finally run successfully for your tests `#include`ing .c files. Of course, we're eager for developers like yourself to get away from this and use the new Partials instead. Perhaps this is the release that first allows you to start doing that…

**Comment by mkarlesky** [comment]
I'm going to (finally!) close this issue. [Ceedling Partials](https://throwtheswitch.github.io/Ceedling/1.1.0/testing-guide/partials/) are here and ready to be exercised. Prereleases are posted with a bit more to go before an official release. The project README is updated, and the Changelog is receiving revisions. More documentation and testing and final issues (outside of Partials) to go before a final release.
